src/compose.ts

```diff
@@ -0,0 +1,228 @@
+				c.networks = _.mapValues(networks, normalizeNetwork);
+			}
+
+			// Normalise volumes
+			if (c.volumes) {
+				const volumes: Dict<Volume | null> = c.volumes;
+				c.volumes = _.mapValues(volumes, normalizeVolume);
+			}
+
+			return c;
+		}
+	}
+}
+
+function normalizeService(service: Service, serviceNames: string[]): Service {
+	if (!service.image && !service.build) {
+		throw new ValidationError('You must specify either an image or a build');
+	}
+
+	if (service.build) {
+		if (_.isString(service.build)) {
+			service.build = { context: service.build };
+		}
+		if (service.build.args) {
+			service.build.args = normalizeKeyValuePairs(service.build.args);
+		}
+		if (service.build.labels) {
+			service.build.labels = normalizeKeyValuePairs(service.build.labels);
+		}
+	}
+
+	if (service.depends_on) {
+		if (_.isArray(service.depends_on)) {
+			if (_.uniq(service.depends_on).length !== service.depends_on.length) {
+				throw new ValidationError('Service dependencies must be unique');
+			}
+			service.depends_on = _(service.depends_on)
+				.map((dep) => {
+					return [ dep, { condition: 'service_started' } ];
+				})
+				.fromPairs()
+				.value();
+		}
+		_.keys(service.depends_on).forEach((dep) => {
+			if (!_.includes(serviceNames, dep)) {
+				throw new ValidationError(`Unknown service dependency: ${dep}`);
+			}
+		});
+	}
+
+	if (service.environment) {
+		service.environment = normalizeKeyValuePairs(service.environment);
+	}
+
+	if (service.extra_hosts) {
+		service.extra_hosts = normalizeKeyValuePairs(service.extra_hosts, ':');
+	}
+
+	if (service.labels) {
+		service.labels = normalizeKeyValuePairs(service.labels);
+	}
+
+	return service;
+}
+
+function normalizeNetwork(network?: Network): Network | null {
+	if (!network) {
+		return null;
+	}
+
+	if (network.labels) {
+		network.labels = normalizeKeyValuePairs(network.labels);
+	}
+
+	return network;
+}
+
+function normalizeVolume(volume?: Volume): Volume | null {
+	if (!volume) {
+		return null;
+	}
+
+	if (volume.labels) {
+		volume.labels = normalizeKeyValuePairs(volume.labels);
+	}
+
+	return volume;
+}
+
+/**
+ * Parses a composition and returns a list of image descriptors that can
+ * be used to pull or build a service image. The given composition version
+ * must be equal to `DEFAULT_SCHEMA_VERSION`, or an exception is thrown.
+ * Normalise the composition before passing it to this function.
+ */
+export function parse(c: Composition): ImageDescriptor[] {
+	if (c.version !== DEFAULT_SCHEMA_VERSION) {
+		throw new Error('Unsupported composition version');
+	}
+	return _.toPairs(c.services).map(([ name, service ]) => {
+		return createImageDescriptor(name, service);
+	});
+}
+
+function createImageDescriptor(serviceName: string, service: Service): ImageDescriptor {
+	if (service.image && !service.build) {
+		return { serviceName, image: service.image };
+	}
+
+	if (!service.build) {
+		throw new Error('should not get here');
```

Comment:
Maybe type this error to `InconsistentParseState` or something so that's it's easier to look up in sentry etc?